form-validation/src/Form.js

```diff
@@ -0,0 +1,177 @@
+  }
+
+  handleSubmit(event) {
+    event.preventDefault();
+    this.clearErrors();
+    if(this.validateFields(event)) {
+      this.resetState();
+    }
+  }
+
+  resetState() {
+    this.setState(this.constructor.initialState);
+  }
+
+  clearErrors() {
+    this.setState({
+      errors: {}
+    });
+  }
+
+  validateFields(event) {
+    let isValid = true
+    for(const element of event.target.elements) {
+      if(this.validatePresence(element)) {
+        isValid = false
+        this.addPresenceValidationError(element);
+      } else if(element.name === 'mail') {
+        isValid = this.validateEmailFormat(element);
+      } else if(element.name === 'homepage') {
+        isValid = this.validateURLFormat(element);
+      } else if(element.name === 'aboutMe') {
+        isValid = this.validateLength(element);
+      }
+    }
+    return isValid;
+  }
+
+  validatePresence(element) {
+    return (!element.value || (element.type === 'checkbox' && !element.checked) || element.selectedIndex <= 0);
+  }
+
+  addPresenceValidationError(element) {
+    this.setState((prevState) => ({
+      errors: {
+        ...prevState.errors,
+        [element.name]: `${element.name} not present`
+      }
+    }));
+  }
+
+  validateEmailFormat(element) {
+    var email = element.value.trim();
+    const emailRegEx = /^[_a-zA-Z0-9]+(\.[_a-zA-Z0-9]+)*@[a-z0-9]+(\.[a-zA-Z0-9]+)*(\.[a-z]{2,4})$/;
+    let validEmail = emailRegEx.test(email)
+    if(!validEmail) {
+      this.setState((prevState) => ({
+        errors: {
+          ...prevState.errors,
+          [element.name]: `${element.name} has invalid format`
+        }
+      }));
+    }
+    return validEmail;
+  };
+
+  validateURLFormat(element) {
+    var url = element.value.trim();
+    const urlRegEx = /^(((http[s]?)|(ftp)):\/\/)?(www\.)?[a-zA-Z0-9\.\-]+\.[a-zA-Z]{2,}[\/]?/;
+    let validURL = urlRegEx.test(url)
+    if(!validURL) {
+      this.setState((prevState) => ({
+        errors: {
+          ...prevState.errors,
+          [element.name]: `${element.name} has invalid format`
+        }
+      }));
+    }
+
+    return validURL;
+  };
+
+  validateLength(element) {
+    const maxLength = 50;
+    let validLength = (element.value.length >= maxLength)
+    if(!validLength) {
+      this.setState((prevState) => ({
+        errors: {
+          ...prevState.errors,
+          [element.name]: `${element.name} should have at least ${maxLength} characters`
+        }
+      }));
+    }
+
+    return validLength;
+  };
+
+  render() {
+    const { fields, errors } = this.state
+    return (
+      <div className='container'>
+        <h2 className='text-center header'>Registration Form</h2>
+        <form className='form' name='form1' onSubmit={this.handleSubmit}>
+          <label className='inputLabels font-bold' htmlFor='loginId'>Login Id</label>
+          <input className='inputs' type='text' id='loginId' name='loginId' value={fields.loginId} onChange={this.handleInputChange} /><br />
```

Comment:
this input & errors tag can also become another components, let's add the label ltag in it as well instead of creating another component.